### DISCUSSION TITLE
2 questions on customing caption of figure

### DISCUSSION THREAD

**Discussion by retsyo** [typst/typst#1013] [General]
1. I have read https://typst.app/docs/reference/meta/figure/ now the caption is rendered as `Figure 1: A curious figure`. What if `Figure 1 A curious figure` or `Figure 1 - A curious figure` is need?

2. What if `Figure 4-1 A curious figure` is need? Where `4` means this in chapter 4, and `1` means the first figure in this chapter

<pre>
# chap 1 soemthing

# chap 2 soemthing
+----------------+
| this is a figure  |
+----------------+
Figure 2-1 A curious figure

# chap 3 soemthing

# chap 4 soemthing
+----------------+
| this is a figure  |
+----------------+

Figure 4-1 A curious figure

</pre>  

thanks

**Comment by LaurenzV** [comment]
Does this do what you want?

```
#set heading(numbering: "1")

#show figure: it => align(center)[
  #it.body
  #it.supplement #counter(heading).display()-#counter(figure).display(it.numbering) #it.caption
]

#show heading: it => [
 #counter(figure).update(0)
 #it
]

= Chapter One

#figure(
  image("glacier.jpg", width: 80%),
  caption: [A curious figure.]
) <glacier>

#figure(
  image("glacier.jpg", width: 80%),
  caption: [A curious figure.]
) <glacier>

#figure(
  image("glacier.jpg", width: 80%),
  caption: [A curious figure.]
) <glacier>

 
= Chapter Two

#figure(
  image("glacier.jpg", width: 80%),
  caption: [A curious figure.]
) <glacier>

#figure(
  image("glacier.jpg", width: 80%),
  caption: [A curious figure.]
) <glacier>

#figure(
  image("glacier.jpg", width: 80%),
  caption: [A curious figure.]
) <glacier>

= Chapter Three

#figure(
  image("glacier.jpg", width: 80%),
  caption: [A curious figure.]
) <glacier>
```

**Reply by retsyo** [reply]
not exactly, there is some differences as the image says
<pre>
#set heading(numbering: "1.1")

#show figure: it => align(center)[
  #it.body
  #it.supplement #counter(heading).display()-#counter(figure).display(it.numbering) #it.caption
]

#show heading: it => [
 #counter(figure).update(0)
 #it
]

= Chapter One

in @glacier

#figure(
  image("img/glacier.png", width: 30%),
  caption: [A curious figure.]
) <glacier>


== subheading of chap 1

in @glacier2

#figure(
  image("img/glacier.png", width: 30%),
  caption: [A curious figure.]
) <glacier2>


#figure(
  image("img/glacier.png", width: 30%),
  caption: [A curious figure.]
) <glacier3>


= Chapter Two

#figure(
  image("img/glacier.png", width: 30%),
  caption: [A curious figure.]
) <glacier4>

#figure(
  image("img/glacier.png", width: 30%),
  caption: [A curious figure.]
) <glacier5>

#figure(
  image("img/glacier.png", width: 30%),
  caption: [A curious figure.]
) <glacier6>

= Chapter Three

#figure(
  image("img/glacier.png", width: 30%),
  caption: [A curious figure.]
) <glacier7>
</pre>


![numbering_figure](https://user-images.githubusercontent.com/7960913/235383628-343c2ec2-9b69-44ad-91cc-70b5ea0e8adc.png)

**Reply by LaurenzV** [reply]
Ah, I see. Here is a version that should do that (however, there seems to be a small bug, namely that if you reference to a picture under a different heading, it doesn't seem to work properly. See for example the second reference, which is referenced as "Figure 1-7" instead of "Figure 3-1". But I _think_ this is a bug in Typst, so I can't really fix this right now. But I hope that still helps you:

```
#set heading(numbering: "1.1")

#show figure: it => align(center)[
  #locate(loc => [
    #it.body
    #it.supplement #counter(figure).display(it.numbering) #it.caption
  ])
]

#show heading.where(level: 1): it => [
 #counter(figure).update(0)
 #it
]

#set figure(numbering: it => {
  locate(loc => {
    [#counter(heading).at(loc).at(0)-#it]
  })
})

= Chapter One

in @glacier you can see.... And in @glacier3 you can see...

#figure(
  image("glacier.jpg", width: 30%),
  caption: [A curious figure.]
) <glacier>


== subheading of chap 1

in @glacier2

#figure(
  image("glacier.jpg", width: 30%),
  caption: [A curious figure.]
) <glacier2>


#figure(
  image("glacier.jpg", width: 30%),
  caption: [A curious figure.]
) 


= Chapter Two

#figure(
  image("glacier.jpg", width: 30%),
  caption: [A curious figure.]
) 

#figure(
  image("glacier.jpg", width: 30%),
  caption: [A curious figure.]
) 

#figure(
  image("glacier.jpg", width: 30%),
  caption: [A curious figure.]
) 

= Chapter Three

#figure(
  image("glacier.jpg", width: 30%),
  caption: [A curious figure.]
) <glacier3>
```

<img width="895" alt="image" src="https://user-images.githubusercontent.com/47084093/235416814-d5be99fb-7280-45a8-aa90-18f3aa061d42.png">

**Comment by retsyo** [comment]
btw, sometimes, the title of a figure and the figure itself run across 2 pages, can typst handle this situation, for example, by shrink the image a little shrink the distance between title and figure, or at least show warning for further investigation

thanks

![numbering_figure](https://user-images.githubusercontent.com/7960913/235384085-8c5c63f8-e0f2-4ee1-8c98-3b9aed2c0a8a.png)

**Reply by LaurenzV** [reply]
You can wrap everything in a `#block(breakable: false)[]`, and then the caption and image should always stay on the same page.

**Comment by retsyo** [comment]
nice, thank you for your codes
